[Do Not Merge] Adding "light wallet server" that is mymonero compatible

vtnerd opened the pull request:
Another [do not merge] :/ . This is in a solid "beta+" code - people should expect this to work reasonably well in its current state. Anybody testing this branch is encouraged to put bug reports or suggestions into this PR.

This patch is too large to be reviewed on its own. Instead pieces will be broken up so that various parts can be discussed on their own. Eventually the last chunk will be this PR, and the do not merge tag will be removed. This PR **will** receive force-pushes to keep the eventual log history somewhat pristine (and it will involve several smaller commits via other PRs anyway). A separate [repo/branch](https://github.com/mymonero/hostedwallet-server-staging/tree/feature/light_wallet_server) will be identical to what goes into mainline, except history will never be rewritten. Some additional unit tests + documentation will be written as the PRs are being opened. 

----------------------------

This provides a mymonero compatible server that uses LMDB and adds no additional dependencies to the project. For now - there is a server that scans txes in the background while an epee server provides the latest data in LMDB over the mymonero REST api. An optional, off by default, exchange rate retrieval from cryptocompare.com is also provided.

This also provides a separate admin utility. The utility can dump the DB state, change some limited account values, approve pending account creation/import requests, and manually add accounts - all while the server is running. The server will automatically detect changes via LMDB, and continue processing.

My biggest criticism - the ideal situation would be to have this in a separate repo, but I think its difficult until the internal parts of the system can be used effectively as a git submodule.

anonimal commented on src/device/device_default.hpp:
```diff
@@ -39,6 +39,18 @@ namespace hw {
+                Unfortunately, "cryptonote_core", "ringct", and "wallet"
+                libraries all depend on the device library creating a cyclic
+                dependency. None of those libraries should be dependent on
+                the hardware abstraction, it should only be dependent on them.
```
>None of those libraries should be dependent on the hardware abstraction, it should only be dependent on them.

:+1:

dternyak commented:
@vtnerd First off, this is awesome, and I solute you for making the effort to get a lot of the magic running MyMonero out there into the wider Monero community.

I wanted to jump in and see if you could give a status update on this PR. 

> This patch is too large to be reviewed on its own. Instead pieces will be broken up so that various parts can be discussed on their own.

What are you thinking in terms of getting the first partial piece of the changes here into a mergeable PR? 

Is this something anyone else can help with beyond review?

FFS? 😁

dternyak commented:
Aha, I just found https://github.com/monero-project/meta/pull/272, which answers a lot of my own questions.

Hopefully that's also useful for onlookers.

vtnerd commented:
There are also two PRs that have been merged - [PR 4371](https://github.com/monero-project/monero/pull/4371) and [PR 4209](https://github.com/monero-project/monero/pull/4209). The current conflicts are related to those files. I will look into rebasing shortly to reduce the diff.

I'm not sure what you can do at the moment, other than to provide feedback on the patches.

vtnerd commented:
Force pushed an update:
  - PR is now updated to latest master revision
  - Random output selection is now done in server process, and closely mirrors wallet2 behavior
    - MyMonero REST API does not send output selection, so choosing outputs around fork is not possible and a small privacy leak to `monerod` occurs. Some `\TODO`s have been left in code.
  - Account lookup is now done by spend public only - allows for the possible extensions for ZMQCurve only authentication
  - Added support for new `generated_locally` flag from REST clients
  - Log level and confirm external bind flags finally work
  - CORs whitelist selection from command line is now possible
  - Several bug fixes

vtnerd commented:
Also, I wanted to mention that the DB is unlikely to change, unless a reviewer notices something serious. So anyone wishing to run/test this server is encouraged to do so, because migration for DB changes will be provided (upon request only pre-merge).

gutenye commented:
@dternyak @anonimal What's the current plan to merge this or timeline to add light wallet API support? The [specification](https://github.com/monero-project/meta/pull/287) is almost finished.

Light wallet does not work anymore after `get_random_outs` RPC is removed in the latest version.

vtnerd commented:
> @dternyak @anonimal What's the current plan to merge this or timeline to add light wallet API support? The specification is almost finished.

I don't think either are responsible for this being merged. One of the dependent reviews (LMDB util stuff) is currently held up in review. There are likely 1-2 other PRs before I remove the `[DO NOT MERGE]` tag.

> Light wallet does not work anymore after get_random_outs RPC is removed in the latest version.

This PR/patch should work with a recent `monerod` built from master or from the 0.13.x branch. It no longer requires `get_random_outs` RPC.

moneromooo-monero commented:
How much work would it be to make this into its own repo, similar to how the GUI does ? It seems very self contained (ie,mostly new files in new directories).

vtnerd commented:
This uses lots of internal headers/functions. So either use the entire source tree to grab internal libraries (like openmonero) or attempt a library type stub here that can be linked against. Both are pretty rough. Better than a library stub specific for this app might be the ominous "librarize" step that was once on the Monero roadmap, but that is even more work.

So I would have to investigate a design for a library stub specific for this task and then decide whether to do this or the internal method that openmonero uses. I'd probably go with the openmonero approach and deal with the constant breaks from the parent repo. Its likely difficult to make a library without being able to use `cryptonote::transaction`, which requires exporting a lot.

moneromooo-monero commented:
FWIW, if the code is in the monero tree, there will be as many constant breaks, except I will get to be the one to fix it.

vtnerd commented:
Fixing the breaks is hopefully done by the person making the breaking change. Often times the issue isn't that some interface changed and needs to be updated, but that the entire behavior changed and the logic is wrapped so tightly into other parts of the codebase that its difficult to re-use or even copy-and-paste the code. Wallet2 output selection is a great example - much of that code could/should be in a separate functions in a header that is used by Wallet2.

I started some re-usability with this patch here; the wallet selection code is not _identical_ (mymonero REST-API doesn't send output ids, so output selection via forking cannot be done), but otherwise the selection algorithm can be re-used by Openmonero without much difficulty. With this patch wallet2/light_wallet_server also share the coinbase filtering/searching logic too for instance.

It might be best to rewrite this patch such that there is no dependency on this project at all. The trickiest part would matching the custom serialization format for the transaction hash prefix calculation, but everything else can probably be re-written into a reusable lib based on tweetnacl (to start anyway).

vtnerd commented:
[Changes that were force pushed](https://github.com/monero-project/monero/compare/aa2da7a..7e9c905):
* Rebased to latest master HEAD.
* Updated scanner to support new "mask"/amount encoding.
* Fixed `ZMQ_LINGER` bug that could cause indefinite "hanging" during shutdown if ZMQ-RPC server was not available.

dternyak commented:
It would be excellent to get this in. So far we haven't seen alternative to MyMonero light wallet providers, and having support for a light wallet server in `monerod` would help to change this.

moneromooo-monero commented:
Then someone who wants this should review it.

dternyak commented:
Totally fair @moneromooo-monero - I would myself if I was familiar enough with the codebase to. Just wanted to drop my two cents in that I think this work would be extremely valuable to include if possible.

vtnerd commented:
@ndorf pointed out the this merge was bad. I hadn't pushed out what was sitting on my laptop somehow. Pushed that out hours ago now.

There will be update some after the [LMDB PR](https://github.com/monero-project/monero/pull/4460) gets merged. It looks like the db schema will be changing to help fix a bug, let me know if you are using this patch in a meaningful way and I will provide an upgrade program.

Jasonhcwong commented:
I checkout the PR and compiled.
But it produced a monero daemon of version 0.13.
Is there a version of light wallet server that will work with monero daemon v0.14.x.x ?

Jasonhcwong commented:
> Totally fair @moneromooo-monero - I would myself if I was familiar enough with the codebase to. Just wanted to drop my two cents in that I think this work would be extremely valuable to include if possible.

Agree that this work is valuable.
I hope that this PR will be merge soon.
I wish I know the codebase more.

Jasonhcwong commented on src/light_wallet_server/scanner.cpp:
```diff
@@ -0,0 +1,756 @@
+            const auto sleep_time = std::min(wait, std::chrono::nanoseconds{interval});
+            const auto start = std::chrono::steady_clock::now();
+            while (scanner::is_running() && (std::chrono::steady_clock::now() - start) < wait)
+                boost::this_thread::sleep_for(boost::chrono::milliseconds{sleep_time.count()});
```
I tested this on my machine Ubuntu 18.04 on x64 platform, it actually sleep for 100 000 times longer than wanted.
With the following code to print real sleep time before the while loop:
`
const boost::chrono::milliseconds real_sleep_time = boost::chrono::milliseconds{sleep_time.count()};            MINFO("real_sleep_time(ms): " << real_sleep_time.count());
`
Suggest change it to :
`
std::this_thread::sleep_for(sleep_time);
`

vtnerd replied:
I missed this comment somehow. The bug is that this is in nanosecods not milliseconds, so its off by 1000. Will push out a fix soon. Glad to see someone testing/using.

Jasonhcwong commented:
Bug report:
Mymonero iOS app crash after a newly created account received a payment.

iOS app version: 1.1.9
Server platform: Ubuntu 18.04, x86_64
Monero daemon: v0.14.0.0-release

To reproduce the crash:
1. prepare clean installation of Mymonero app without any existing accounts.
2. prepare a clean installation of monero light wallet server database.
3. create a new account on Mymonero app and set to connect to your own monero light wallet server
4. on the monero light wallet server, accept the create request from Mymonero iOS app.
5. wait for some time and the lightwallet server will start scanning blocks(once a new block is found) for your account.
6. wait for some time to make sure that everything works with a newly created EMPTY account
7. send some fund to the newly create account and wait for confirmation and lightwallet server processed the block.
8. Mymonero iOS crashed as soon as it received the new "get_address_info" and "get_address_txs"(with a transaction) msg from lightwallet server.

Log:
I printed the json message at the server side:
Before any payment is made:
// get_address_info
{
  "locked_funds": "0",
  "total_received": "0",
  "total_sent": "0",
  "scanned_height": 1814180,
  "scanned_block_height": 1814180,
  "start_height": 1814180,
  "transaction_height": 1814180,
  "blockchain_height": 1814180,
  "spent_outputs": []
}
// get_address_txs
{
  "total_received": "0",
  "scanned_height": 1814180,
  "scanned_block_height": 1814180,
  "start_height": 1814180,
  "transaction_height": 1814180,
  "blockchain_height": 1814180,
  "transactions": []
}

After the payment is confirmed:
// get_address_info
{
  "locked_funds": "0",
  "total_received": "99931240000",
  "total_sent": "0",
  "scanned_height": 1814184,
  "scanned_block_height": 1814184,
  "start_height": 1814180,
  "transaction_height": 1814184,
  "blockchain_height": 1814184,
  "spent_outputs": []
}
// get_address_txs
{
  "total_received": "99931240000",
  "scanned_height": 1814184,
  "scanned_block_height": 1814184,
  "start_height": 1814180,
  "transaction_height": 1814184,
  "blockchain_height": 1814184,
  "transactions": [
    {
      "id": 0,
      "hash": "e5ef3d522cf848c303e7dfbf6d2b9b48664c20f91ab630930110f53380abc2d3",
      "timestamp": "2019-04-16T08:28:33.0-00:00",
      "total_received": "99931240000",
      "total_sent": "0",
      "unlock_time": 0,
      "height": 1814183,
      "payment_id": "0000000000000000",
      "coinbase": false,
      "mempool": false,
      "mixin": 10,
      "spent_outputs": []
    }
  ]
}

It seems that the "get_address_txs" with a transaction made Mymonero iOS app crash.

paulshapiro commented:
hey @Jasonhcwong 

Could you gather the crash log and make an issue on github.com/mymonero/mymonero-app-ios? It's probably some discrepancy with this lightwallet server's response format but I didn't spot anything immediately in what you sent, so the crash log (once symbolicated) will be a help. Actually, since you're probably running a debug build, can you engage the debugger and show me a screenshot of where it crashes? Else I'll need your dSYM file.

Thanks
Paul

vtnerd commented:
Force pushed a lot of updates:

- Rebased to recent master commit
- Server now uses most recent output selection algorithm
- SSL support was added to the REST server, with custom key/cert just like other monero applications
- Multiple REST servers can be specified on a command line, i.e. `--rest-server http://127.0.0.1:8080 --rest-server https://0.0.0.0:8443` is now possible. The internal non-http server allows for Tor/i2p hidden service forwarding (Tor/i2p handle the encryption/auth) without having to run a second server process
- Several important bug fixes:
  - Similar to simplewallet, receiving coinbase outputs now always ignore the commitment 
  - Receiving multiple outputs to same address now filtered (required DB change)

No one notified me that they were running this, so I didn't write any code to update the DB. If this is an issue contact me directly. Users **should** expect this PR to remain valid with mainline, so running locally should not be problematic.

At least one PR as a result of merging recent changes will likely be done, possibly two.

dternyak commented:
@stoffu Any possibility of you reviewing this? It looks like there's demand for a light wallet server that'll run optionally in monerod from this thread alone. Users would gain a lot of privacy benefits from configuring wallets like Edge or MyMonero to use a light wallet server that they run, so I can see this having a lot of real world privacy implications.

moneroexamples commented:
Hi,

I'm looking at the code and I can't locate handling subaddresses. How are they handled in the light wallet server? How does the server store the outputs identified to belong to subaddresses? How is this information transferred to the frontend?

Also I don't understand how `import_request` is supposed to work? https://github.com/vtnerd/monero/blob/light_wallet_server/src/light_wallet_server/rest_server.cpp#L894

I see it has `import_fee` field as in the [API docs](https://github.com/monero-project/meta/blob/c139cb39ea56441fe7cb7ff13b74fe34b394f2f5/api/lightwallet_rest.md#import_request). But what about `payment_address` and legacy `payment_id` as in the API specification? If there is no `payment_address` and legacy `payment_id`, how the client should provide information on where to send the fee required?

vtnerd commented:
The API does not support subaddresses and neither does this light wallet server. The load this puts on the server is a bit funky. I thought we discussed this at some point @moneroexamples , am I thinking of someone else?

The process for approving addresses in this app is currently manual. The admin process can be used concurrently with the running server process to make changes. The output of the admin process is always JSON, and the inputs are positional process arguments. The structure was designed such that the CLI admin commands should be usable/adaptable to a REST or ZMQ/JSON-RPC system.

The `import_fee` is hard-coded to zero and `payment_address` and `payment_id` are omitted from the response. I left it manual for now, because it seems unlikely that the `import_request` logic would be included as part of the project since the needs are likely to vary. The system was designed around ZMQ a bit (more than a bit actually), so the likely extension would be to use ZMQ-PUB/SUB to push import request events, and accept DB changes through the same channel.

moneroexamples commented:
Thanks for rely. 

Yes, we discussed sub-addresses [already](https://github.com/monero-project/monero/pull/2056#issuecomment-488542664). And although it puts some extra load on the server, I think that the benefits of sub-addresses out weight the computational cost.  Without sub-addresses, the light wallet users are forced to use their primary address for all txs, which is not good for privacy. Also such light-wallet can't be used even as a view-only wallet, since it does not recognize outputs addressed to sub-addresses and spendings of these outputs.

Thus, I don't know what is the current status of this PR?. Are sub-addresseses planned to be added in near future? This obviously would require changes to the rest API to convey information about sub-addresses to the clients. 

The way `import_request` works is interesting, because I don't see how it relates to the way mymonero  works which has non-zero fee.  If the aim of this PR is to provide mymonero compatible server to monero daemon, I think it should include ability to import external wallets for a fee. The default can be zero, but if someone wants to open their light-wallet node for others, a fee could be required.

ndorf commented:
This PR contains something potentially quite valuable: a reusable output selection implementation. I was easily able to link the relevant modules into another C++ program. Only three files are needed from this PR, along with a couple more from `crypto` and `common`:

        wallet/output_selection.cpp
        light_wallet_server/error.cpp
        light_wallet_server/random_outputs.cpp

Since the official selection algorithm in wallet2 is not easily reusable, third-party wallet implementors must create their own implementations that must not only be correct, but indistinguishable from the official one and each other. That seems unrealistic to put it mildly, and issues like #5712 show the sad, inevitable results (some wallet is using locked outputs as ring members).

A reusable official implementation would be much better, and it looks like a part of this PR gets us much of the way there. This one actually differs slightly from the wallet2 one (because it doesn't know the true spent output), so that's at least one thing that would need to be addressed. Nonetheless, isn't this worth exploring further?

vtnerd commented:
Got them build bots green!:
 * Fixed some conflicts with master
 * Fixed libcxx 3.9 build (older XCode)
 * Fixed Windows build
 * Fixed 32-bit Linux build
 * Changed enum to Monero styling - @moneromooo-monero commented on this in a different review of mine -> camel case to underscores
 * Changed `error::kNoSuchAccount` -> `error::account_not_found`.

fuwa0529 commented:
A light server could make sense for now, but since the price of SSD has been free falling, it wouldn't be completely unthinkable that within two years, having a copy of the monero blockchain wherever it's needed is cheap enough for most users.

The need of having a daemon processing txs for a wallet would probably be obsolete.

There will be users that refuse to run their own nodes whatsoever, but then they do not contribute to the strength of the network at all. Shouldn't devs make their lives harder rather then easier?

vtnerd commented:
This type of server is useful even for personal use - scanning on a phone is still a major pain, and people frequently have more CPU/bandwidth/power on home machines.

paulshapiro commented:
Can we merge this?

moneromooo-monero commented:
My opinion is that it belongs in a separate repo since it's a separate program, uses a separate blockchain db, offers a different API, but if someone else like smooth, hyc, etc OK it (and review it) it'll go in.

paulshapiro commented:
@moneromooo-monero my view is that the lightwallet server offers an alternative API to wallet2 and RPC - both of which are distinct APIs, themselves. In the not too distant future, the lightwallet server will support hybrid scanning mode. At that time, wallet programmers have little to no reason to spend (probably more than) twice the amount of time coding to two different APIs just to get both fullwallet and lightwallet - this will allow them to code to a single API and get both, which means the LWS could well become the de facto API for wallet integrations. Is there an argument against this? Curious to know. Secondly, in order to move this LWS out to a separate repo, we need to factor the entire Monero core codebase to make it embeddable. That's obviously a non-starter. So many things from this repository belong in a separate repository. So, so many things. And so, so many things absolutely don't belong in wallet2 whatsoever. At least this code is excellently written and factored and actually co-exists properly. I don't see it as fair to single put this PR and penalize us just because new datastructures are introduced here. The API is not new or different. wallet2 has had merged lightwallet support for ages, and we've also included the LWS API spec standard in the Monero meta repo, so I really don't think it's right to categorize this as something that somehow belongs less here.

paulshapiro commented:
There are strong reasons to include it side-by-side with daemon, as well.

paulshapiro commented:
By the way, as to arguments against a singular API for full and light – the shape/format of the API changes significantly to allow for that, so we're using the chance to design fixes for many things. So the argument against the LWS API for replacing existing ones that it's not a perfect API isn't applicable here - certain advancements needed for full/hybrid that we have design solution suggestions to introduce shortly.

paulshapiro commented:
@hyc curious to know your view too btw

paulshapiro commented:
Hey so one more thing that occurs to me is that people *should* be able to download these things together… yes, as mentioned before, we want people to run their own LWS servers rather than have to put load on our hosting, but beyond that, people ought to be able to download the software bundled together because it doesn't make sense for a third party to maintain some separate repo or build and direct people to download that, while also directing them to download Monero. The LWS + API are already used by multiple wallets in the space. Having to rebase this forever keeps it in a state that people can't / don't practically access it and run the code. This work creates more opportunity to safely improve Monero usability and having it in the Monero code repo would only bring more visibility to that. I think it's clear the community wants it, and the contribution is not malicious (or at least it can be argued to be significantly less introductory of code complexity than other things of less importance to Monero privacy but which *are* being merged), so I hope we can find a way to get this done.

stoffu commented:
@paulshapiro 
> people ought to be able to download the software bundled together because it doesn't make sense for a third party to maintain some separate repo or build and direct people to download that, while also directing them to download Monero

The GUI is considered to be part of Monero's official releases and is downloaded by many users, so I think the LWS can take the similar approach (i.e. maintained as a separate repository under monero-project and included in the official releases).

fluffypony commented:
@stoffu the GUIs a bad example - we want people *already running full nodes* to be able to use light wallets and/or offer light wallet support in a random pool (which is currently done by remote node). By having it as a separate piece of software that just adds extra friction for people to use it. It also means that guides to running a Monero mobile wallet securely will involve third-party software that needs to be trusted to bundle the daemon correctly, which is also an issue. I would support either the current strategy, or at worst it’s own repo on the monero-project GitHub.

stoffu commented:
@fluffypony 
> the GUIs a bad example - we want people already running full nodes

I don't see how the GUI is a bad example - the GUI users are also supposed (or at least recommended) to run the daemon locally.

> By having it as a separate piece of software that just adds extra friction for people to use it. It also means that guides to running a Monero mobile wallet securely will involve third-party software that needs to be trusted to bundle the daemon correctly, which is also an issue.

I don't see any difference here either. The LWS can just be bundled together with the GUI, or can be packaged separately and linked in the official Downloads page. The package can of course come with a user guide for the LWS, just like the GUI. I wouldn't call the LWS third-party software if it's hosted under monero-project and released and maintained by the core team.

moneromooo-monero commented:
I get that keeping it in a separate repo means you'll have to maintain it, but that was kinda my point earlier: if it is in the monero repo itself, *I* will get to have to maintain a parallel implementation of part of monero, because every time I change something that deals with the blockchain, it'll likely break that code, and you wouldn't want me to push something which breaks the build, right ? So I'll get to maintain that new separate thing (and also anyone who happens to modify whatever part of the code this PR relies on). Even if I make it build, I have no plans to run a light wallet node just to test my PRs to monero and ensure that I also fix it for those changes. It'll likely just end up either wasting my time every time I change blockchain code, or falling by the wayside like the 0MQ RPC unless vtnerd (or you, or anyone else who wants to) maintains it (in the second case). And if it is so, it's better in a separate repo IMHO, because it'll be as much work, but it won't fall on me as some kind of neverending yak shaving task :)

But as I said, if one other "main" person like smooth or hyc or stoffu wants it in, it goes in. I'll also add pony in there, and it seems he'd rather have it in. So if it's reviewed seriously (you can do it too), I'm not blocking it.

paulshapiro commented:
That this PR (which makes so much sense and is so desired) is at the threshold of unmergability, to me, indicates more about the degree of entropy of the Monero codebase than the scope of the addition.

paulshapiro commented:
I don't think the maintenance of the entire Monero codebase can fall on only your shoulders when that is to include both the daemon and the wallet code. It can be argued the daemon itself does not belong in Monero core, too, but rather the impls of the essential processes and functions.

paulshapiro commented:
I'd say as well we have a priorities problem too if the MMS was merged to the dangerous wallet2 and this excellently written code cannot be.

moneromooo-monero commented:
Stop fucking around. I did not say it cannot be. I said SEVERAL times I'm not blocking it, I'm just not helping it either.

moneromooo-monero commented:
Though I do agree mms doesn't really belong there either, but hey.

moneromooo-monero commented:
Is one of you guys maintaining the existing light wallet code in wallet2 ? AFAIK it's bitrotted. Would you maintain it if this PR went in ?

paulshapiro commented:
@moneromooo-monero you're right, the LW client code in wallet2 is badly bitrotted - it hasn't been maintained, and it ought to crash outright due to a condition in an assert having been inverted. I don't believe anyone's actually using the LW code in wallet2… and whatever value we could have gotten from it being in C++ is now obtained indirectly from the C++ implementation of the LW client in mymonero-core-cpp… which we're pleased to maintain… so how about we deprecate the wallet2 LW client codepaths and datastructures in wallet2 in favor of this PR?

moneromooo-monero commented:
Would you be PRing that new light wallet code to the monero repo too ?
I have no objections to the light wallet code being removed if noone uses it.

paulshapiro commented:
I would love for that code to be considered for PR. But such a PR involves a factoring of wallet2, I think.

+1 thank you!

paulshapiro commented:
I mean it would be a partial factoring but the main question is how we converge / collate / coalesce the transaction construction code, and then, it's mostly a matter of wallet2 calling the pure functions I've extracted from it into mymonero-core-cpp

moneromooo-monero commented:
It was a question, not a request. Mainly because I see it as unwanted, and merging the daemon part might mean you come back later and ask for the wallet part to be merged too :)

paulshapiro commented:
I think you might view most of mymonero-core-cpp as not strictly describable as "that new light wallet code" - since much of is straight lifted from Monero, actually. So if you mean only the lightwallet-specific code in mymonero-core-cpp is not desired, then I'd like to mention that I think I've always advocated for separation of repos somehow…… and maybe what we'd do is have a Monero core of some kind which has all the scanning code and Monero wallet mode code.. including all daemon underlying impl… and this would basically be the union of libmonerocore and the hypothetical libwallet …… then I guess we'd have the daemon and LWS in another repo, probably with simplewallet… then the GUI in a third repo? That way the GUI and simplewallet would code to the same underlying corewallet API. If so, whatever new lightwallet code there is would simply have unit tests added for it in that libmonerocore repo for its additions to the corewallet API. Just some thoughts. I don't mind PRing the removal of the existing dead LW code in wallet2 but am unsure exactly when I'll get to it - probably a week or so

vtnerd commented:
I've been looking at how to take this code into a separate repo - whether it is possible to have a Cmake file within the repo that other projects can import. This would be useful for this PR, openmonero, mymonero, and @jtgrassie mining pool. These other projects are already having to "hook" into the monero as-if it were designed to be a library, when it has not.

Certain parts, like the device code (for ledger, etc.), have now "infected" several chunks of code due to a (what seems to be at quick glance) a poor design. There's also a weirdness between cryptonote_core / cryptonote_basic - some stuff from core should perhaps be moved to basic. You end up having to chop out functions manually to reduce dependencies or just accept the entire `monerod` codebase shoved into your project (including its logging system).

All of this makes putting this PR into a separate repo complicated because some re-architecting will be necessary. But multiple projects would likely benefit. The trouble is that I'm not sure how viable this is long-term - it will be easy to muck things up to accelerate some PR getting merged in the future. I can't even give a rough estimate after ~45 mins, but its fairly close to being easier to do something closer to what @endogenic has been suggesting for wallet libs - just copy/fork and refactor it to barebones and use that instead and then slowly introduce those new libs into monerod.

-------

On the topic of wallet2 + light-wallet code. I'd have to spend time looking at wallet2 again, but my thoughts have typically been that light-wallet code was really hacked into wallet2. The design would _probably_ work better if it were inverted such that wallet2 had a generic interface for requesting possible spends (i.e. conform to the light-wallet interface). If the scanning is local the implementation just returns the results locally immediately, etc. One issue is that I don't recall build breaks or test failures in this area.

vtnerd commented:
Later this evening I will look at the opposite, see how stuff can be just copied/forked out (including using Paul's existing lib code).

moneromooo-monero commented:
I'm fine with some factoring, as long as it doesn't become wholesale rewriting (in which case it'd belong as a separate "wallet3" lib, which monero could switch to at some point after it's well tested).

moneromooo-monero commented:
Though of course this wallet3, if it was made, would risk going the way of the light wallet code or the 0mq RPC, which are bitrotting because noone maintains them, so you'd have to keep it up to date regularly for it to work in practice.

vtnerd commented:
I don't see how 0mq is bitrotting exactly - it still has the same features/info from before. It just hasn't kept up with the other API for new fields. The LWS server still worked as expected, and would other software likely.

moneromooo-monero commented:
That's what I meant, yes. I guess bitrotting was not the best word for it then :)

vtnerd commented:
There are four high-level ways forward:
 * Keep this existing PR as-is
 * Move to another repo and fork dependencies into that repo
 * Move to another repo and refactor ("librarize") so that external projects can use some critical parts
 * Just hook in as-is and let static linker drop unused function or let the unused portions sit on disk (its not terabytes of ASM, etc). This is what openmonero and the mining pool implementation are basically doing. Kind of tricky since you have to include monero's logger, etc.

FWIW, "librarize" monero was on the roadmap at some point. 

-------------

This PR depends on these components within this repo:
 * Epee HTTP server. Librarize not recommended. Either fork or use [nodejs C http parser](https://github.com/nodejs/http-parser). Or use beast. Or write one from scratch. This isn't really terrifying me.
 * `src/crypto` - librarize is the _most_ doable but has an epee dependency that may need to be cleaned up. Forking also an option. Re-writing from ref10 baseline or donna64 is doable but not necessarily recommended since exporting shouldn't be _too_ painful. Most of the PoW code isn't really useful either (miners "do their own thing"); may want to split libs.
 * `src/ringct` - librarize or fork. There is no re-writing this realistically. Has a dependency on `src/crypto` and `src/common` - not sure how easy it is to drop the latter.  An unlisted epee depency (log) should be droppable with macro-magic.
 * `src/cryptonote_[basic|core]`. These two are similar - there are handful of useful functions spread throughout these libs that are combined with things not needed. Re-writing will be difficult at times - getting a transaction hash or prefix_hash for instance will be fairly painful rewrite. Other parts (base58 stuff) will be somewhat easy to fork/rewrite. Librarize will be a bit more painful as a bunch of functions will have to be rolled out but it seems plausible at glance.

I have no recommendations at the moment.

vtnerd commented:
Force pushed changes:
 * Rebased to master, fixing build
 * Fixed sleeping bug.

paulshapiro commented:
@vtnerd total support

ndorf commented:
https://github.com/mymonero/monero-core-custom could be a good starting point. This library contains a small subset of upstream Monero code for use in wallet implementations. The same content, but with Monero itself as a git ancestor, is available here: https://github.com/mymonero/monero/tree/core-custom-master (might be worth using `git remote add` to add this one to your existing Monero clone, as almost all of the 300MB+ repo is identical.)

Presently, the files themselves differ significantly from their upstream versions, but most or all of those differences are just removing or disabling extra functionality. It should be fairly straightforward to clean that up using one or more preprocessor macros.

Anyway, this is already sufficient to use the output selection functionality of this PR. Some additional files would be needed for the entire server (starting with the ZMQ message types, I think).

woodser commented:
FWIW I have started looking into refactoring wallet2 and I expect I will continue, so please consider me available to help.

The first thing that jumped out is that the light wallet is mostly an independent wallet within wallet2.  Like @vtnerd said, ideally Monero offers a single interface which full and light wallets implement separately.  The full wallet should probably offer a base class which minimizes application dependencies (e.g. using an interface to make network requests, not writing to disk, etc) in order to be run in different application environments (e.g. JavaScript for the browser or Go for OpenBazaar).

I figured a good way to go about this is to first refactor the underlying wallet2 assets into e.g. wallet2_base, wallet2_full, and wallet2_light, then introduce a wallet3 interface and wrapper implementations which the GUI, RPC, and application developers can use long term.

I don't have an opinion on where the light wallet assets should live now, but I do support the light wallet being officially supported by Monero Core.  I would expect the light wallet to be alongside the full wallet and I expect it would be easiest to maintain the light wallet server if it is "close" to the other Core code, especially during these refactors.  Otherwise realistically if our most active contributors maintain one and not the other because they're in separate repositories, wouldn't they quickly fall out of sync?

paulshapiro commented:
@woodser that was the route I took in factoring wallet2 (think we may have discussed) - wallet3_base, wallet3_light, etc (see https://github.com/mymonero/mymonero-core-cpp--old - which is abandoned) but given the upcoming hybrid mode, having classes separated like that will cause whichever code hangs onto the reference to the wallet object instance to have to rebuild the instance of the wallet if they want to switch to another wallet type (in the soon to come hybrid mode). That's not the worst in the world but it probably makes sense to incorporate that kind of switching logic into a wallet class - and I don't think it's implicitly a good idea to say it belongs in a wallet manager class - not just because that's a route down the same black hole. I think a solution will be to not keep the implementations in those classes at all - but rather in monero core itself in the form of pure functions - then the actual wallet implementation itself could be trivially and potentially preferably implemented by whoever is making their own wallet software - but it would also expand options we have in writing whichever one(s) is/are used by Monero core. Invariably we're going to have some separation of code by repo / library … and I think the only way around that is for Monero to commit to having a particular API at each release and then deprecate that API like other projects. The API can come with unit tests and be negotiated in the same way as existing APIs are added.. based on needs, and such. Keeping them in separate repositories is an example of the definition of an increase in complexity but it provides many benefits. All in all, if the wallet code ended up in core, that doesn't seem like an issue to me, but the more important thing is the stability of the API imo, as @vtnerd also mentions.

paulshapiro commented:
At the very least, we don't need to do anything to wallet2 except extract things from it so that those implementations can be shared... wallet2 can remain exactly the same externally - and its functions would simply end up calling / proxying into an extraction of their implementations. I think that as an initial goal would provide the majority of the benefits everyone's looking for.

vtnerd commented:
This will still be moved to another repo - in the meantime I pushed a rebase + merge and did some basic tests against it. Anyone running this early release should be good for the fork.

vtnerd commented:
Bringing this up again. The strategy to get this moving:

1. Add a new RPC to ZMQ that accepts raw bytes for forwarding a tx. The HTTP RPC has this. The server will no longer need custom Monero serialization or ringct stuff. All of that is pushed onto the clients of the server.
2. Make `monero-project/monero` an `external/` folder in a new repo. Only the `src/crypto` and `contrib/epee` should be a require/mandatory dependency. But `src/crypto` has several other dependencies (randomx ahhhh).
3. Replace the custom JSON stuff in this PR with the approach in [issue 6406](https://github.com/monero-project/monero/issues/6406). The `BOOST_PP` code will be dropped to reduce complexity further. The resulting code will be closer to that in ZMQ-JSON, and faster since it supports DOMless reading/writing. All of this has been written _already_, including the DOMless code. This should reduce review friction, compared to whats in here now.
4. Attempt to somehow PR this in stages to some `moneroo-project` repo with reviews.
5. Add support for the wallet ASM.
6. Replace epee http with something more "standard". This will reduce breakages from changes in `monero-project/monero`, and (probably) result in some performance improvement.
7. Add donna64 to replace usage of `monero-project/monero/src/crypto` ... ? Aggressive, but removes the difficulties of the dependency madness. Nothing from `monero-project/monero` should need to be pulled in after a few functions are copied.

Anyway 1-4 is doable and gets people using this.

ndorf commented:
> 1. Add a new RPC to ZMQ that accepts raw bytes for forwarding a tx. The HTTP RPC has this. The server will no longer need custom Monero serialization or ringct stuff. All of that is pushed onto the clients of the server.

This was done in #5207, `send_raw_tx_hex` is the new call.

vtnerd commented:
I even commented on that PR. And I updated the lookup table with that entry too. Well, hitting middle age I guess.

vtnerd commented:
[monero-lws](https://www.github.com/vtnerd/monero-lws) is now in a separate repo and functional.
